snapcraft/internal/lifecycle.py

```diff
@@ -112,6 +118,39 @@ def execute(step, project_options, part_names=None):
             'type': config.data.get('type', '')}
 
 
+def _setup_core(deb_arch):
+    core_path = os.path.join(os.path.sep, 'snap', 'core', 'current')
+    if os.path.exists(core_path):
+        logger.debug('{!r} already exists, skipping core setup'.format(
+            core_path))
+        return
+    core_revision = snapcraft.get_latest_revision('core')
+    snap_cache = SnapCache(project_name='core')
+
+    # Clean up old stuff first
+    snap_cache.prune(keep_revision=core_revision)
+
+    # And try to get this revision.
+    core_snap = snap_cache.get(snap_name='core', revision=core_revision)
+    if not core_snap:
+        with tempfile.TemporaryDirectory() as d:
+            snap_name = os.path.join(d, 'core_stable_{}.snap'.format(deb_arch))
+            snapcraft.download('core', 'stable', snap_name, deb_arch)
+            # FIXME there is a chance of incorrecly caching the wrong
+            # revision here.
+            snap_cache.cache(snap_name, revision=core_revision)
+    # This should now be a hit.
+    core_snap = snap_cache.get(snap_name='core', revision=core_revision)
```

Comment:
maybe, you could simplify this with a try catch. Make get raise an exception instead of return None.

```
try:
    core_snap = cache.get...
except NotFoundInCache:
    core_snap = download  # Raise if fails to download.
    save_to_cache(core_snap)
```

I don't see why you need to call core_snap twice, and check if not core_snap twice.